core/src/main/java/org/jruby/RubyStruct.java

```diff
@@ -327,11 +348,45 @@ private void checkSize(int length) {
         }
     }
 
+    private static void checkNewInstanceKeywordArguments(RubyHash args) {
+        Ruby runtime = args.getRuntime();
+        RubySymbol keywordInitSym = runtime.newSymbol("keyword_init");
+        if (args.size() > 1) {
+            List unknownKeywords = Arrays.stream(args.keys().toJavaArray())
+                .filter(k -> !k.equals(keywordInitSym))
+                .map(IRubyObject::asJavaString)
+                .collect(Collectors.toList());
+            throw runtime.newArgumentError("unknown keyword" +
+                (unknownKeywords.size() > 1 ? "s: " : ": ") +
+                unknownKeywords.stream().collect(Collectors.joining(", ")));
+        }
+    }
+
+    private void checkKeywordArguments(RubyHash args) {
+        if (args.size() != values.length) {
+            RubyArray members = __member__();
+            String unknownKeywords = Arrays.stream(args.keys().toJavaArray())
+                .filter(o -> !members.contains(o))
+                .map(IRubyObject::asJavaString)
+                .collect(Collectors.joining(", "));
+            throw getRuntime().newArgumentError("unknown keywords: " + unknownKeywords);
+        }
+    }
+
```

Comment:
I realised I was reimplementing functionality already available in JRuby with these checks. I'll update this PR soon to address that.